### PR TITLE
Local/benchmark test

### PR DESCRIPTION
**Pull request by jinsooihm**
## Description

Added benchmark feature to `circle.ci`. 
When pushed to git, print and save benchmark data in` circle.ci`.
Exit with `1` if the new commit is slower than the origin/master by 0.7 times.

## Further work
The input sizes to run benchmarks on should be calibrated in a way that `circle.ci` should have enough resource to run. 
Some of the functions can be improved in organization.
Currently, benchmark takes very long on `circle.ci`.
More documentation can be added.
Only some of `freud` module benchmarks (`locality`, `density`) have been added yet.

## Types of changes

- [x] New feature (non-breaking change which adds or improves functionality)

## Checklist:

- [x] My code follows the code style of this project.
- [x] I have read the [**CONTRIBUTING**](https://github.com/glotzerlab/freud/blob/master/CONTRIBUTING.md) document.

### CODE REVIEW

File: .circleci/config.yml
Comment (bdice) [inline .circleci/config.yml:95]: Is there a reason why you don't just call `export PYTHONPATH=$PYTHONPATH:.` directly?

File: .circleci/config.yml
Comment (bdice) [inline .circleci/config.yml]: `origin`

File: .circleci/config.yml
Comment (bdice) [inline .circleci/config.yml:121]: The current output on CircleCI for "Uploading artifacts" says:
```
Uploading /home/ci/ci/freud/benchmarks/reports to benchmarks-reports
No artifact files found at /home/ci/ci/freud/benchmarks/reports
```
I think `path` has to point to a specific file - please refer to the CircleCI docs and figure out how to upload all the result files generated by the benchmarks.

File: benchmarks/benchmark.py
Comment (bdice) [inline benchmarks/benchmark.py]: Just put `self` in for `test_wrapper` and remove line 43.

File: benchmarks/benchmark.py
Comment (bdice) [inline benchmarks/benchmark.py]: initialize

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: I'd like to remove the thread count from the signature of every `run` method, too -- we could use the environment variable `BENCHMARK_NPROC` just like we're already using `BENCHMARK_NPROC_INCREMENT`, to make the implementation of all this cleaner.

Alternatively we could refactor this to take an integer and limit to that number of threads, instead of making it specific to CircleCI. The CircleCI config can specify 2 threads via the command line with something like `-j 2` or `--threads 2` instead of `-c`.

Ask @vyasr for his opinion.

File: benchmarks/benchmark.py
Comment (bdice) [inline benchmarks/benchmark.py]: See my other comment below - I'd like to use something like `limit_threads=P` or an environment variable `BENCHMARK_NPROC` instead of `on_circleci=False`.

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: In `freud`, we use Google-style docstrings (put the function documentation below the `def` and use the format below).
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

Comment (bdice) [reply benchmarks/benchmarker.py]: I'd like to rename this function to something like `run_benchmarks`.

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: This description of adding benchmarks should end up in the freud documentation, right below the Unit Tests in the developer guide. https://freud.readthedocs.io/en/stable/design.html#unit-tests

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: initialization

**Review by bdice — CHANGES_REQUESTED** [review]
Nice work, @jinsoo960. I have requested some changes in the PR.

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py:408]: extra space

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: Is there a way you could write this and the printing logic above as a single function?

File: benchmarks/benchmarker.py
Comment (bdice) [inline benchmarks/benchmarker.py]: Add this as a function docstring in complete sentences.

File: .circleci/config.yml
Comment (bdice) [inline .circleci/config.yml:116]: I'd really like to put the comparison into its own task so that it appears as a separate section in the CircleCI output. Something called `Compare benchmarks`, right below `Run benchmarks`.

image

Comment (bdice) [reply .circleci/config.yml:116]: This would also clarify that the benchmarks succeeded but their _comparison_ failed (e.g. by being slow).

File: .circleci/config.yml
Comment (jinsooihm) [reply .circleci/config.yml:95]: https://circleci.com/docs/2.0/env-vars/#setting-an-environment-variable-in-a-shell-command seems to recommend doing so.

File: benchmarks/benchmark.py
Comment (vyasr) [inline benchmarks/benchmark.py]: Can we keep the explicit `(object)` inheritance for now? As long as we're supporting Python2.7 I don't want to run into any unexpected MRO surprises.

File: benchmarks/benchmark_box_ParticleBuffer.py
Comment (vyasr) [inline benchmarks/benchmark_box_ParticleBuffer.py]: Can we remove?

File: benchmarks/benchmarker.py
Comment (vyasr) [inline benchmarks/benchmarker.py]: Can we change the signature to `print_stats=True` and save ourselves the duplication above?

File: benchmarks/benchmark_environment_BondOrder.py
Comment (vyasr) [inline benchmarks/benchmark_environment_BondOrder.py]: Remove

File: benchmarks/benchmarker.py
Comment (vyasr) [reply benchmarks/benchmarker.py]: I'm also happy with adding default arguments to anything else (like `number`) if they're pretty standard across our benchmarks, but `print_stats` is the obvious one.

File: benchmarks/benchmark_order_SolLiq.py
Comment (vyasr) [inline benchmarks/benchmark_order_SolLiq.py]: Is there a reason the last argument is "hidden" (with an underscore)?

File: benchmarks/util.py
Comment (vyasr) [inline benchmarks/util.py]: If you're already using rowan in a benchmark, you can probably use it here as well. I would probably remove this function entirely.

File: doc/source/design.rst
Comment (vyasr) [inline doc/source/design.rst]: Need to swap these

**Review by vyasr — COMMENTED** [review]
Almost there! Some minor cleanup to do then this is ready to go I think.

**Comment by jinsooihm** [comment]
Reply to 

> Is there a reason the last argument is "hidden" (with an underscore)?

The reason is that `flake8` does not like having `I` as a variable name.

**Comment by vyasr** [comment]
Got it, `flake8` is pretty strict about that. If you can't come up with a better name, then the better solution to renaming the variable `_I` is just tell `flake8` to ignore that line with `#noqa`. For an example see [order.py](https://github.com/glotzerlab/freud/blob/master/freud/order.pyx#L1552).

File: .circleci/config.yml
Comment (bdice) [inline .circleci/config.yml:141]: @jinsoo960 Sorry, I was unclear earlier. It looks like we're running the benchmarks twice (once to see if the benchmarks pass and again with comparison). You already did what I actually wanted when you split the `benchmark` step into `benchmark` and `comparison`.

We're running more CI than we really need to in this case. Just remove the `comparisons` job and add `- *comparison` as a step to this job.

File: benchmarks/benchmark.py
Comment (bdice) [inline benchmarks/benchmark.py]: Please convert these comments to a Google docstring. (and so on, below)

File: benchmarks/benchmark.py
Comment (bdice) [inline benchmarks/benchmark.py]: Please convert these comments to a Google docstring.

File: benchmarks/benchmark_density_LocalDensity.py
Comment (bdice) [inline benchmarks/benchmark_density_LocalDensity.py:20]: Optional: We could create a helper function `random_points(box, N)` in the `util` file for all our benchmarks, that generates random points between `[0, 1]` in each dimension (accounting for the `box` dimensionality, with z=0 for 2D boxes), and then uses `box.makeCoordinates`.

Comment (bdice) [reply benchmarks/benchmark_density_LocalDensity.py:20]: I would prefer to initialize the box in the setup function, not the run function.

File: benchmarks/benchmark_order_LocalQlNear.py
Comment (bdice) [inline benchmarks/benchmark_order_LocalQlNear.py]: The `_I` parameter here is actually not an `I` "eye" at all - it's supposed to be a lowercase `l` "ell." flake8 will also be mad if you use a lowercase `l`, since it's easy to mix up with a capital `I` (as we see). Please rename `_I` to `sph_l` for clarity, that's the convention we've adopted in a few places.

(same for other Steinhardt-variant classes, RotationalAutocorrelation, etc.)

File: doc/source/design.rst
Comment (bdice) [inline doc/source/design.rst]: Missing something like `from benchmarker import run_benchmarks`

Please check this to be sure this example file matches the implemented design. To make it simpler, we could just include a real benchmark with [the `.. literalinclude::` directive](https://www.sphinx-doc.org/en/1.5/markup/code.html#directive-literalinclude) and say "The following code is an example benchmark for the `freud.density.RDF` module," etc.

File: doc/source/design.rst
Comment (bdice) [inline doc/source/design.rst]: The latter sentence might be clearer if you said, "Benchmarks are run as a part of continuous integration, with performance comparisons between the current commit and the `master` branch."

**Review by bdice — CHANGES_REQUESTED** [review]
Some comments for improvement, then this should be ready to go.

**Comment by bdice** [comment]
@jinsoo960 One more thing I forgot to mention - there's a folder of benchmarks you labeled "to be removed." If that code can be superseded by proper benchmarks for each of those modules, just remove those files completely.

**Comment by vyasr** [comment]
@bdice I think all your comments have been addressed, but can you check to be sure?

**Comment by bdice** [comment]
@vyasr @jinsoo960 I made some changes, I approve it in the current state after all tests pass (I don't think I broke anything but I didn't test it locally so let's read the CI benchmark reports one more time before merging).

**Review by bdice — APPROVED** [review]

**Comment by vyasr** [comment]
New changes look fine, and all tests pass.

**Pull request merged**
